Add type converters suffixed by `OrNull` and `OrThrow` for `StrictlyNegativeInt`

o-korpi opened the pull request:
Close #149.

o-korpi committed:
feat(#149): add experimental builders for StrictlyNegativeInt

New builders: toStrictlyNegativeIntOrNull and toStrictlyNegativeIntOrThrow.
These are not implemented yet.

o-korpi committed:
feat(#149): implement the toStrictlyNegativeIntOrNull function

o-korpi committed:
feat(#149): implement the toStrictlyNegativeIntOrThrow function

LVMVRQUXL commented:
Thank you @o-korpi for your first contribution to this project! 🙏🏽
I will review your request very soon. 👀

LVMVRQUXL commented on src/commonMain/kotlin/kotools/types/number/StrictlyNegativeInt.kt:
```diff
@@ -22,6 +22,56 @@ import kotlin.jvm.JvmInline
+ */
+@ExperimentalNumberApi
+@ExperimentalSinceKotoolsTypes("4.3")
+public fun Number.toStrictlyNegativeIntOrNull(): StrictlyNegativeInt? =
```
We will release this function in the version [4.4.0](https://github.com/kotools/types/milestone/7).

```suggestion
@ExperimentalSinceKotoolsTypes("4.4")
public fun Number.toStrictlyNegativeIntOrNull(): StrictlyNegativeInt? =
```

LVMVRQUXL commented on src/commonMain/kotlin/kotools/types/number/StrictlyNegativeInt.kt:
```diff
@@ -22,6 +22,56 @@ import kotlin.jvm.JvmInline
+
+/**
+ * Returns this number as a [StrictlyNegativeInt], which may involve rounding
+ * or truncation, or throws [IllegalArgumentException] if this number is
```
There is a typo.

```suggestion
 * or truncation, or throws an [IllegalArgumentException] if this number is
```

LVMVRQUXL commented on src/commonMain/kotlin/kotools/types/number/StrictlyNegativeInt.kt:
```diff
@@ -22,6 +22,56 @@ import kotlin.jvm.JvmInline
+ */
+@ExperimentalNumberApi
+@ExperimentalSinceKotoolsTypes("4.3")
+public fun Number.toStrictlyNegativeIntOrThrow(): StrictlyNegativeInt =
```
We will release this function in the version [4.4.0](https://github.com/kotools/types/milestone/7).

```suggestion
@ExperimentalSinceKotoolsTypes("4.4")
public fun Number.toStrictlyNegativeIntOrThrow(): StrictlyNegativeInt =
```

LVMVRQUXL commented on src/commonTest/kotlin/kotools/types/number/StrictlyNegativeIntTest.kt:
```diff
@@ -71,6 +72,39 @@ class StrictlyNegativeIntTest {
+    fun number_toStrictlyNegativeIntOrNull_should_not_be_null_with_a_strictly_negative_Int() {
+        val number: Number = StrictlyNegativeInt.random().toInt()
+        val result: StrictlyNegativeInt? = number.toStrictlyNegativeIntOrNull()
+        result?.toInt() shouldEqual number
```
We should check first that the `result` shouldn't be `null` by using the [`shouldBeNotNull`](https://github.com/kotools/types/blob/39c92ca9cd9b0373f5613b7e5da6950e278715ba/src/commonTest/kotlin/kotools/types/Assertions.kt#L20) function.

```suggestion
        result.shouldBeNotNull().toInt() shouldEqual number
```

LVMVRQUXL commented on src/commonTest/kotlin/kotools/types/number/StrictlyNegativeIntTest.kt:
```diff
@@ -71,6 +72,39 @@ class StrictlyNegativeIntTest {
+    fun number_toStrictlyNegativeIntOrNull_should_be_null_with_a_positive_Int() {
+        val number: Number = PositiveInt.random().toInt()
+        val result: StrictlyNegativeInt? = number.toStrictlyNegativeIntOrNull()
+        assertNull(result)
```
We should align assertions by using the [`shouldBeNull`](https://github.com/kotools/types/blob/39c92ca9cd9b0373f5613b7e5da6950e278715ba/src/commonTest/kotlin/kotools/types/Assertions.kt#L22) function instead.

```suggestion
        result.shouldBeNull()
```

LVMVRQUXL commented on src/commonTest/kotlin/kotools/types/number/StrictlyNegativeIntTest.kt:
```diff
@@ -71,6 +72,39 @@ class StrictlyNegativeIntTest {
+    @Test
+    fun number_toStrictlyNegativeIntOrThrow_should_pass_with_a_negative_Int() {
+        val number: Number = StrictlyNegativeInt.random().toInt()
+        number.toStrictlyNegativeIntOrThrow().toInt() shouldEqual number
```
Declaring a `result` variable is useful for checking the return type of the tested function.

```suggestion
        val result: StrictlyNegativeInt = number.toStrictlyNegativeIntOrThrow()
        result.toInt() shouldEqual number
```

LVMVRQUXL commented on src/commonTest/kotlin/kotools/types/number/StrictlyNegativeIntTest.kt:
```diff
@@ -71,6 +72,39 @@ class StrictlyNegativeIntTest {
+        val exception: IllegalArgumentException = assertFailsWith<IllegalArgumentException> {
+            number.toStrictlyNegativeIntOrThrow()
+        }
+        exception.shouldHaveAMessage()
```
We should align assertions by using the [`shouldFailWithIllegalArgumentException`](https://github.com/kotools/types/blob/39c92ca9cd9b0373f5613b7e5da6950e278715ba/src/commonTest/kotlin/kotools/types/Assertions.kt#L24) function instead.

```suggestion
        number.shouldFailWithIllegalArgumentException {
            toStrictlyNegativeIntOrThrow()
        }.shouldHaveAMessage()
```

LVMVRQUXL reviewed (changes requested):
@o-korpi I will work on your branch for integrating your changes to this project.
Here's the checklist that I'm addressing for integrating those:
- [ ] Rebase the feature branch on the `main` branch of this project while signing commits of the feature branch.
- [x] Resolve the conversations below.

o-korpi committed:
fix(#149): correct typo in KDoc of toStrictlyNegativeIntOrThrow

o-korpi committed:
fix(#149): correct version for toStrictlyNegativeIntOrThrow

o-korpi committed:
fix(#149): correct version for toStrictlyNegativeIntOrNull

o-korpi committed:
test(#149): update tests of toStrictlyNegativeIntOrNull function

- Add `shouldNotBeNull` before checking equality.
- Use `shouldBeNull` instead of `assertNull`.

o-korpi committed:
test(#149): update tests of toStrictlyNegativeIntOrThrow function

- Declare variable to check return type of function.
- Use the `shouldFailWithIllegalArgumentException` function instead
  of `assertFailsWith<IllegalArgumentException>`.

LVMVRQUXL commented on src/commonTest/kotlin/kotools/types/number/StrictlyNegativeIntTest.kt:
```diff
@@ -5,17 +5,14 @@ import kotlinx.serialization.SerializationException
 import kotlinx.serialization.decodeFromString
 import kotlinx.serialization.encodeToString
 import kotlinx.serialization.json.Json
+import kotools.types.*
```
We try not to use wildcard imports in the library.

LVMVRQUXL reviewed (changes requested):
Thank you @o-korpi for updating your branch with the changes requested above.
Please resolve the conversations below for integrating your work.

o-korpi committed:
refactor(#149): replace wildcard import by explicit imports

LVMVRQUXL reviewed (changes requested):
Thank you @o-korpi for updating your branch with the changes requested above.
Here's the checklist that I'm addressing for integrating your work:
- [x] Rebase the feature branch on the main branch of this project while signing commits of the feature branch.

LVMVRQUXL reviewed (approved):
Congratulations @o-korpi, this PR can be merged! 🎉
I will merge your branch later when the checks will succeed.
Thank you again for your contribution!
